pkg/api/vm/formatter.go

```diff
@@ -1,25 +1,93 @@
 package vm
 
 import (
+	"strings"
+
 	"github.com/rancher/apiserver/pkg/types"
+	ctlkubevirtv1alpha3 "github.com/rancher/harvester/pkg/generated/controllers/kubevirt.io/v1alpha3"
 	"github.com/rancher/steve/pkg/resources/common"
+
+	v1alpha3 "kubevirt.io/client-go/api/v1alpha3"
+	cdiv1 "kubevirt.io/containerized-data-importer/pkg/apis/core/v1alpha1"
 )
 
 const (
-	startVM   = "startVM"
-	stopVM    = "stopVM"
-	restartVM = "restartVM"
+	startVM      = "startVM"
+	stopVM       = "stopVM"
+	restartVM    = "restartVM"
+	unmountCDRom = "unmountCDRom"
 )
 
-func formatter(request *types.APIRequest, resource *types.RawResource) {
+type vmFormater struct {
+	vmCache ctlkubevirtv1alpha3.VirtualMachineCache
+}
+
+func (vf *vmFormater) format(request *types.APIRequest, resource *types.RawResource) {
 	common.Formatter(request, resource)
 	// reset resource actions, because action map already be set when add actions handler,
 	// but current framework can't support use formatter to remove key from action map
 	resource.Actions = make(map[string]string, 1)
 	if resource.APIObject.Data().Bool("spec", "running") {
 		resource.AddAction(request, stopVM)
 		resource.AddAction(request, restartVM)
+
+		data := resource.APIObject.Data()
+		vmName := data.String("metadata", "name")
+		vmNamespace := data.String("metadata", "namespace")
+
+		vm, err := vf.vmCache.Get(vmNamespace, vmName)
+		if err == nil {
+			copyObj := vm.DeepCopy()
+			if cdRomExisted := unmountCDRomFromVM(copyObj); cdRomExisted {
+				resource.AddAction(request, unmountCDRom)
+			}
+		}
+
 	} else {
 		resource.AddAction(request, startVM)
 	}
 }
+
+func unmountCDRomFromVM(vm *v1alpha3.VirtualMachine) bool {
+	var cdRomExisted bool
+	var dvName, volumeName string
+	if !vmReady.IsTrue(vm) {
+		return cdRomExisted
+	}
+
+	var dvs []cdiv1.DataVolume
+	for _, v := range vm.Spec.DataVolumeTemplates {
+		if v.Spec.Source.HTTP != nil && strings.HasSuffix(v.Spec.Source.HTTP.URL, "iso") {
```

Comment:
This is not the proper way to find a cdrom, we should check the cdrom field: https://kubevirt.io/user-guide/#/creation/disks-and-volumes?id=cdrom